### PR TITLE
Push file to s3 bucket

### PR DESCRIPTION
**Pull request by tuanngocnguyen**
Required changes in main plugin for step form validation

https://github.com/catalyst/moodle-tool_lifecycle/pull/5/files

S3 feature is optional and requires local_aws

### CODE REVIEW

**Comment by dmitriim** [comment]
Getting following error when adding a step:
`Notice: Trying to access array offset on value of type null in /var/www/monash39/admin/tool/lcbackupcoursestep/classes/lifecycle/step.php on line 380`

image

**Comment by dmitriim** [comment]
And then after saving following:

Notice: Trying to access array offset on value of type null in /var/www/monash39/admin/tool/lcbackupcoursestep/classes/lifecycle/step.php on line 380

Notice: Undefined index: usesdkcreds in /var/www/monash39/admin/tool/lcbackupcoursestep/classes/lifecycle/step.php on line 346
[Skip to main content](https://monash39.localhost/admin/tool/lifecycle/editelement.php#maincontent)

File: classes/s3/helper.php [hunk 1]
```diff
@@ -0,0 +1,139 @@
...
+ */
+class helper {
+
+    /**
+     * Create a new S3 client.
+     *
+     * @param array $settings amazon s3 settings.
+     * @return \Aws\S3\S3Client
+     */
+    public static function create_client($settings) {
```
Comment (dmitriim) [inline classes/s3/helper.php:41]: if we expect settings as an array, let's cast array

File: classes/s3/helper.php [hunk 2]
```diff
@@ -0,0 +1,139 @@
...
+        return new \Aws\S3\S3Client($options);
+    }
+
+    /**
+     * Test connection to S3 bucket.
+     *
+     * @param array $settings amazon s3 settings.
+     * @return stdClass
+     */
+    public static function check_connection($settings): stdClass {
```
Comment (dmitriim) [inline classes/s3/helper.php:78]: if we expect settings as an array, let's cast array

File: classes/s3/helper.php [hunk 3]
```diff
@@ -0,0 +1,139 @@
...
+        if (!$usesdkcreds) {
+            $options['credentials'] = [
+                'key' => $settings['s3_key'],
+                'secret' => $settings['s3_secret'],
+            ];
+        }
+
+        // Proxy.
+        if ($settings['s3_useproxy']) {
+            $options['http'] = ['proxy' => \local_aws\local\aws_helper::get_proxy_string()];
```
Comment (dmitriim) [inline classes/s3/helper.php:59]: do we need to add a dependency on local_aws?

File: classes/s3/helper.php [hunk 4]
```diff
@@ -0,0 +1,139 @@
...
+        try {
+            $client = self::create_client($settings);
+            $client->headBucket(['Bucket' => $settings['s3_bucket']]);
+        } catch (\Aws\S3\Exception\S3Exception $e) {
+            $connection->success = false;
+            $connection->details = $e->getMessage();
+        } catch (\Exception $e) {
+            $connection->success = false;
+            $connection->details = $e->getMessage();
+        }
```
Comment (dmitriim) [inline classes/s3/helper.php:91]: What is the point to have two catches here with the same bodies?

File: classes/s3/helper.php [hunk 5]
```diff
@@ -0,0 +1,139 @@
...
+
+    /**
+     * Upload file to S3 bucket.
+     *
+     * @param int $processid process id.
+     * @param int $instanceid step instance id.
+     * @param int $courseid course id.
+     * @param stored_file $file file to upload.
+     */
+    public static function upload_file($processid, $instanceid, $courseid, $file) {
```
Comment (dmitriim) [inline classes/s3/helper.php:103]: should we type cast parameters here?

File: classes/s3/helper.php (hunk 3 above)
Comment (tuanngocnguyen) [reply classes/s3/helper.php:59]: I want to make the feature optional, so no need for the dependency

File: classes/s3/helper.php (hunk 1 above)
Comment (tuanngocnguyen) [reply classes/s3/helper.php]: I have fixed this.

File: classes/s3/helper.php (hunk 2 above)
Comment (tuanngocnguyen) [reply classes/s3/helper.php]: I have fixed this

File: classes/s3/helper.php (hunk 3 above)
Comment (tuanngocnguyen) [reply classes/s3/helper.php:65]: I have added a soft dependency check

File: classes/s3/helper.php (hunk 4 above)
Comment (tuanngocnguyen) [reply classes/s3/helper.php]: I have combined the catch

File: classes/s3/helper.php (hunk 5 above)
Comment (tuanngocnguyen) [reply classes/s3/helper.php]: I have added the param types

File: classes/s3/helper.php [hunk 6]
```diff
@@ -0,0 +1,142 @@
...
+     * Create a new S3 client.
+     *
+     * @param array $settings amazon s3 settings.
+     * @return \Aws\S3\S3Client
+     */
+    private static function create_client(array $settings) {
+        global $CFG;
+        // Soft dependency on AWS SDK.
+        if (!file_exists($CFG->dirroot . '/local/aws/version.php')) {
+            throw new \moodle_exception('s3_unmet_dependency', 'tool_lcbackupcoursestep');
```
[the remark this answers is not in the document]
Comment (dmitriim) [reply classes/s3/helper.php:45]: Maybe abstract checking for AWS SDK in a method and then fire this method between anything that would user AWS stuff. For 3.9 version you can check for `file_exists($CFG->dirroot . '/local/aws/version.php'` in the future versions (were AWS SDK is part of core) we can just get this method return true or completely remove it.

[the remark this answers is not in the document]
Comment (tuanngocnguyen) [reply classes/s3/helper.php]: Hi @dmitriim 
I have updated the MR with a new method

File: classes/lifecycle/step.php [hunk 7]
```diff
@@ -157,6 +162,15 @@ public function instance_settings() {
             new instance_setting('backup_groups', PARAM_BOOL, true),
             new instance_setting('backup_competencies', PARAM_BOOL, true),
             new instance_setting('backup_contentbankcontent', PARAM_BOOL, true),
+
+            // S3 settings.
```
Comment (dmitriim) [inline classes/lifecycle/step.php:166]: Is it safe to have these setting if dependency is not met?

**Pull request merged**